File: rasiin_healthcare_insurance/rasiin_healthcare_insurance/report/insurance_claims_report/insurance_claims_report.py
# Copyright (c) 2024, Ahmed Ibar and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data

def get_columns(filters):
    if filters.get("detailed"):
        # Columns for the detailed view
        return [
            {"label": "Insurance Company", "fieldname": "insurance_company", "fieldtype": "Link", "options": "Insurance Company", "width": 200},
            {"label": "Hospital Name", "fieldname": "company", "fieldtype": "Link", "options": "Company", "width": 200},
            {"label": "Patient", "fieldname": "patient", "fieldtype": "Data", "width": 150},
            {"label": "Patient Name", "fieldname": "patient_name", "fieldtype": "Data", "width": 200},
            {"label": "Date of Birth", "fieldname": "dob", "fieldtype": "Data", "width": 150},
            {"label": "Policy Number", "fieldname": "policy_number", "fieldtype": "Data", "width": 150},
            {"label": "Date of Service", "fieldname": "invoice_date", "fieldtype": "Data", "width": 150},
            {"label": "Coverage Limit", "fieldname": "coverage_limits", "fieldtype": "Data", "options": "Insurance Company", "width": 150},
            {"label": "Invoice Number", "fieldname": "sales_invoice", "fieldtype": "Link", "options": "Sales Invoice", "width": 200},
            {"label": "Journal Number", "fieldname": "reference_journal", "fieldtype": "Link", "options": "Journal Entry", "width": 200},
            {"label": "Invoice Amount", "fieldname": "total_invoiced", "fieldtype": "Currency", "width": 150},
            {"label": "Patient Amount", "fieldname": "patient_amount", "fieldtype": "Currency", "width": 150},
            {"label": "Insurance Covered Amount", "fieldname": "insurance_amount", "fieldtype": "Currency", "width": 150},
            {"label": "Insurance Paid Amount", "fieldname": "insurance_paid", "fieldtype": "Currency", "width": 150},
            {"label": "Insurance Outstanding Amount", "fieldname": "insurance_outstanding", "fieldtype": "Currency", "width": 150},
        ]
    else:
        # Columns for the summary view
        return [
            # {"label": "Claim Number", "fieldname": "claim_number", "fieldtype": "Link", "options": "Insurance Claim", "width": 200},
            # {"label": "Claim Date", "fieldname": "claimed_date", "fieldtype": "Date", "width": 150},
            {"label": "Insurance Company", "fieldname": "insurance_company", "fieldtype": "Link", "options": "Insurance Company", "width": 200},
            {"label": "Number of Invoices", "fieldname": "total_invoices", "fieldtype": "Number", "width": 150},
            # {"label": "Claim Total Amount", "fieldname": "claim_total", "fieldtype": "Currency", "width": 150},
            {"label": "Insurance Covered Amount", "fieldname": "total_insurance_amount", "fieldtype": "Currency", "width": 150},
            {"label": "Paid Amount", "fieldname": "total_payment_amount", "fieldtype": "Currency", "width": 150},
            {"label": "Outstanding Amount", "fieldname": "insurance_outstanding", "fieldtype": "Currency", "width": 150},
            # {"label": "Claim Status", "fieldname": "claim_status", "fieldtype": "Data", "width": 100},
        ]

def get_data(filters):
    conditions = []
    params = {}

    if filters.get("insurance_company"):
        conditions.append("si.insurance_company = %(insurance_company)s")
        params["insurance_company"] = filters["insurance_company"]
    else:
        # Ensure only invoices with an insurance company are included
        conditions.append("si.insurance_company IS NOT NULL AND si.insurance_company != ''")

    if filters.get("patient"):
        conditions.append("si.patient = %(patient)s")
        params["patient"] = filters["patient"]

    if filters.get("from_date") and filters.get("to_date"):
        conditions.append("si.posting_date BETWEEN %(from_date)s AND %(to_date)s")
        params["from_date"] = filters["from_date"]
        params["to_date"] = filters["to_date"]

    if filters.get("invoice_status"):
        conditions.append("si.status = %(invoice_status)s")
        params["invoice_status"] = filters["invoice_status"]

    where_clause = " AND ".join(conditions)
    if where_clause:
        where_clause = "WHERE si.docstatus = 1 And si.coverage_limits >0 AND " + where_clause
    else:
        where_clause = "WHERE si.docstatus = 1 and si.coverage_limits> 0 "

    if filters.get("detailed"):
        query = f"""
            SELECT 
                si.company AS company,
                si.name AS sales_invoice,
                si.reference_journal AS reference_journal,
                si.posting_date AS invoice_date,
                si.patient AS patient,
                si.patient_name AS patient_name,
                si.insurance_company,
                si.coverage_limits,
                si.grand_total AS total_invoiced,
                si.insurance_coverage_amount AS insurance_amount,
                si.payable_amount AS patient_amount,
                si.patient_paid AS patient_paid,
                si.insurance_paid AS insurance_paid,
                (si.payable_amount - si.patient_paid) AS patient_outstanding,
                (si.insurance_coverage_amount - si.insurance_paid) AS insurance_outstanding,
                si.status AS invoice_status,
                p.dob AS dob,
                si.policy_number AS policy_number

            FROM 
                `tabSales Invoice` si
            LEFT JOIN `tabPatient` p ON p.name = si.patient
            {where_clause}
            ORDER BY 
                p.patient_name ASC
        """
    else:
        query = f"""
            SELECT 
                si.insurance_company,
                COUNT(si.name) AS total_invoices,
                SUM(si.insurance_coverage_amount) AS total_insurance_amount,
                SUM(si.insurance_paid) AS total_payment_amount,
                (SUM(si.insurance_coverage_amount) - SUM(si.insurance_paid)) AS insurance_outstanding
            FROM 
                `tabSales Invoice` si
            {where_clause}
            GROUP BY 
                si.insurance_company
            ORDER BY 
                si.posting_date DESC
        """

    return frappe.db.sql(query, params, as_dict=True)

File: rasiin_healthcare_insurance/rasiin_healthcare_insurance/report/insurance_claims_report/insurance_claims_report.js
// Copyright (c) 2024, Ahmed Ibar and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Insurance Claims Report"] = {
	"filters": [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			reqd: 1
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1
		},
		{
			fieldname: "insurance_company",
			label: __("Insurance Company"),
			fieldtype: "Link",
			options: "Insurance Company"
		},
		{
			fieldname: "patient",
			label: __("Patient"),
			fieldtype: "Link",
			options: "Patient"
		},
		{
			fieldname: "detailed",
			label: __("Detailed View"),
			fieldtype: "Check",
			default: 0
		}
	],
	onload: function(report) {
        report.page.add_inner_button(__("Print Invoices"), function() {
            // Get report data
            frappe.call({
                method: "frappe.desk.query_report.run",
                args: {
                    report_name: "Insurance Claims Report",
                    filters: frappe.query_report.get_filter_values()
                },
                callback: function(response) {
                    let report_data = response.message.result;
                    
                    if (!report_data || report_data.length === 0) {
                        frappe.msgprint(__('No data found to print.'));
                        return;
                    }

                    // Extract all invoice numbers from the report
                    let invoice_numbers = report_data
                        .map(row => row.sales_invoice)
                        .filter(Boolean); // Remove any null or undefined values

                    if (invoice_numbers.length === 0) {
                        frappe.msgprint(__('No valid invoices found to print.'));
                        return;
                    }

                    // Build the URL for bulk printing
                    const url = `/api/method/frappe.utils.print_format.download_multi_pdf?doctype=Sales Invoice`
                        + `&name=${encodeURIComponent(JSON.stringify(invoice_numbers))}`
                        + `&format=Insurance Print Format` // Change to your print format name
                        + `&no_letterhead=0`
                        + `&options=${encodeURIComponent(JSON.stringify({ "page-size": "A4" }))}`;

                    // Open the generated PDF in a new tab
                    window.open(url, "_blank");
                }
            });
        });
    }
};
